### PR TITLE
SDK Function Call API + main class for SDK

### PR DESCRIPTION
**Pull request by drinkcoffee**
This pull request creates a new entry point class for the SDK: net.consensys.gpact.CrosschainProtocols   The idea is that this is the central point class from which all functionality can be accessed. The class allows for registration of new implementations from third party vendors. At present, this class only supports the function call API. Messaging layer classes will be added in a separate PR.

This pull request also creates an API for the function call layer. The example code access functionality using CrosschainProtocols and this API only: no access to implementation classes is needed. As a result of the common API, how to execute a crosschain transaction with the GPACT and SFC protocols has changed slightly. These changes are reflected in the example code.

### CODE REVIEW

**Comment by drinkcoffee** [comment]
When considering this PR, an important class to consider is: CrosschainProtocols. Is the API the approach that everyone is happy with:  https://github.com/ConsenSys/gpact/pull/91/files#diff-50addbac16a7bfb7c1b8d7e0fb74f45176a6c25ce8cc7573639581643f4a72ac

**Comment by drinkcoffee** [comment]
Another thing which I have considered is moving all implementation / non API packages into a sub-package "internal". That is   the packages from change as shown below. The idea would be to make it obvious that the classes should not be used directly, even if a constructor, method or field was public.

In this PR / current                                                   If we had "internal"
net.consensys.gpact.functioncall.common           net.consensys.gpact.functioncall.internal.common 
net.consensys.gpact.functioncall.gpact                net.consensys.gpact.functioncall.internal.gpact
net.consensys.gpact.functioncall.sfc                    net.consensys.gpact.functioncall.internal.
sfc

File: examples/gpact/write/java/src/main/java/net/consensys/gpact/examples/gpact/write/GpactCrosschainWrite.java [hunk 1]
```diff
@@ -95,17 +92,16 @@ public static void main(String[] args) throws Exception {
...
-      boolean success = executionEngine.execute(callGraph, 300);
+      LOG.info("Success: {}", result.isSuccessful());
 
-      LOG.info("Success: {}", success);
-
-      List<BigInteger> callP = new ArrayList<>();
-      callP.add(BigInteger.ZERO);
-      TransactionReceipt txR = executor.getTransationReceipt(callP);
-      bc2ContractBBlockchain.showEvents(txR);
+      //      List<BigInteger> callP = new ArrayList<>();
```
Comment (drinkcoffee) [inline examples/gpact/write/java/src/main/java/net/consensys/gpact/examples/gpact/write/GpactCrosschainWrite.java:101]: TODO: Re-add the code that goes through the transaction receipts.

File: examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java [hunk 2]
```diff
@@ -163,20 +160,19 @@ public static void main(String[] args) throws Exception {
...
+    LOG.info("Success: {}", result.isSuccessful());
 
     bc1TradeWalletBlockchain.showAllTrades();
 
-    List<BigInteger> callP = new ArrayList<>();
-    callP.add(BigInteger.ONE);
-    callP.add(BigInteger.ZERO);
-    TransactionReceipt txR = executor.getTransationReceipt(callP);
-    bc2BusLogicBlockchain.showEvents(txR);
+    //    List<BigInteger> callP = new ArrayList<>();
```
Comment (drinkcoffee) [inline examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java:171]: TODO: Re-add code to check transaction receipts.

**Review by ermyas — APPROVED** [review]

File: sdk/java/src/main/java/net/consensys/gpact/functioncall/CrossControlManagerGroup.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+   *
+   * <p>Configure a blockchain for use by the function call layer.
+   *
+   * @param creds Credentials to be used for signing transactions on the specified blockchain.
+   * @param bcInfo Information about the blocchain: Blockchain id, URL, gas pricing strategy.
+   * @param messageVerification Implementation used to obtain signed events or proofs of events so
+   *     that the events can be used on remote blockchains.
+   * @throws Exception If an issue is encountered deploying the crosschain control contract.
+   */
+  void addBlockchainAndDeployContracts(
```
Comment (ricardolyn) [inline sdk/java/src/main/java/net/consensys/gpact/functioncall/CrossControlManagerGroup.java:27]: I don't see the value if you keep adding contract deployment code to a Java SDK directly. This code should be on deployment scripts that are common in whatever language you have. 

Duplicating this for each SDK is not scalable.

I started a new small project using the GPACT contracts and see how I can set up the the examples read/write using Hardhat+TS.

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+/** Entry point class for the Crosschian Protocol Stack SDK. */
+public class CrosschainProtocols {
+  public static final String GPACT = "GPACT";
+  public static final String SFC = "SFC";
+
+  // Execution engine tpe
+  public static final String SERIAL = "SERIAL";
+  public static final String PARALLEL = "PARALLEL";
+
+  private static final Map<String, Class> functionCallImpls;
```
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java:32]: I had tried to get this to work some days ago, but couldn't work out the syntax. I have solved this now. 
The required syntax is: 
`Class<? extends CrossControlManagerGroup>`
"extends" is the correct key word, even though CrossControlManagerGroup is an interface.

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * Register a crosschain function call protocol implementation. This method can be used to update
+   * the implementation of a protocol by passing in a name that is associated with an existing
+   * implementation.
+   *
+   * @param name Name to associate with an implementation.
+   * @param implClass Protocol implementation class that implements CrossControlManagerGroup.
+   */
+  public static void registerFunctionCallImpl(final String name, final Class implClass) {
```
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java]: See above comment

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+   * Create an instance of a crosschain function call implementation.
+   *
+   * @param implementationName Name of the implementation that must have been previously registered.
+   * @return Implementation of CrosschainFunctionCall
+   * @throws Exception Typically thrown if the implementation has not been registered yet.
+   */
+  public static CrossControlManagerGroup getFunctionCallInstance(final String implementationName)
+      throws Exception {
+    final Class<CrossControlManagerGroup> clazz = functionCallImpls.get(implementationName);
+    return clazz.getDeclaredConstructor().newInstance();
```
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java:67]: That sounds like a good idea.

File: sdk/java/src/main/java/net/consensys/gpact/functioncall/CrossControlManagerGroup.java (hunk 3 above)
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/functioncall/CrossControlManagerGroup.java:27]: I think that different users will want to do things differently. Especially for testing / getting things working initially, I can imagine users wanting to be able to deploy an entire system and run some scenario all in the one executable.  Once the deployment scripts have been created, maybe this API could call to the deployment script.

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * Register a crosschain function call protocol implementation. This method can be used to update
+   * the implementation of a protocol by passing in a name that is associated with an existing
+   * implementation.
+   *
+   * @param name Name to associate with an implementation.
+   * @param implClass Protocol implementation class that implements CrossControlManagerGroup.
+   */
+  public static void registerFunctionCallImpl(final String name, final Class implClass) {
+    if (implClass == null) {
```
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java:51]: done

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import net.consensys.gpact.functioncall.sfc.SimpleCrossControlManagerGroup;
+
+/** Entry point class for the Crosschian Protocol Stack SDK. */
+public class CrosschainProtocols {
+  public static final String GPACT = "GPACT";
+  public static final String SFC = "SFC";
+
+  // Execution engine tpe
+  public static final String SERIAL = "SERIAL";
+  public static final String PARALLEL = "PARALLEL";
```
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java:31]: I think there are going to be multiple execution engines in the future - some defined outside of this SDK but used with the API. As such, I think String is what is needed.

[the remark this answers is not in the document]
Comment (ermyas) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java:31]: Ok, I assumed they were fixed.

File: examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java (hunk 2 above)
Comment (drinkcoffee) [reply examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java:171]: Re-added this

File: examples/gpact/write/java/src/main/java/net/consensys/gpact/examples/gpact/write/GpactCrosschainWrite.java (hunk 1 above)
Comment (drinkcoffee) [reply examples/gpact/write/java/src/main/java/net/consensys/gpact/examples/gpact/write/GpactCrosschainWrite.java]: done

File: examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java (hunk 2 above)
Comment (drinkcoffee) [reply examples/gpact/trade/java/src/main/java/net/consensys/gpact/examples/gpact/trade/Main.java]: done

File: sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (drinkcoffee) [reply sdk/java/src/main/java/net/consensys/gpact/CrosschainProtocols.java]: done

**Pull request merged**